src/provider/provider_coarse.c

```diff
@@ -49,8 +50,256 @@ typedef struct block_t {
+
+    if (!head_node->head) {
+        util_mutex_unlock(head_node->mutex);
+        return NULL;
+    }
+
+    node_t *node = head_node->head;
+    assert(node->prev == NULL);
+    if (node->next) {
+        node->next->prev = NULL;
+    }
+
+    head_node->head = node->next;
+    util_mutex_unlock(head_node->mutex);
+
+    struct block_t *block = node->block;
+    block->node = NULL;
+    free(node);
+
+    return block;
+}
+
+static block_t *node_list_rm(head_node_t *head_node, node_t *node) {
+    assert(head_node);
+    assert(node);
+
+    util_mutex_lock(head_node->mutex);
+
+    if (!head_node->head) {
+        util_mutex_unlock(head_node->mutex);
+        return NULL;
+    }
+
+    if (node == head_node->head) {
+        assert(node->prev == NULL);
+        head_node->head = node->next;
+    }
+
+    node_t *node_next = node->next;
+    node_t *node_prev = node->prev;
+    if (node_next) {
+        node_next->prev = node_prev;
+    }
+    if (node_prev) {
+        node_prev->next = node_next;
+    }
+
+    util_mutex_unlock(head_node->mutex);
+    struct block_t *block = node->block;
+    block->node = NULL;
+    free(node);
+
+    return block;
+}
+
+static int allocated_blocks_add(struct ravl *allocated_blocks, block_t *block) {
+    ravl_data_t data = {(uintptr_t)block->data, block};
+    return ravl_emplace_copy(allocated_blocks, &data);
+}
+
+static block_t *allocated_blocks_rm(struct ravl *allocated_blocks, void *ptr) {
+    ravl_data_t data = {(uintptr_t)ptr, NULL};
+    struct ravl_node *node;
+    node = ravl_find(allocated_blocks, &data, RAVL_PREDICATE_EQUAL);
+    if (!node) {
+        return NULL;
+    }
+    ravl_data_t *node_data = ravl_data(node);
+    if (!node_data) {
+        return NULL;
+    }
+    block_t *block = node_data->value;
+    ravl_remove(allocated_blocks, node);
+    return block;
+}
+
+static int free_blocks_add(struct ravl *free_blocks, block_t *block) {
+    head_node_t *head_node = NULL;
+    int rv;
+
+    ravl_data_t data = {(uintptr_t)block->size, NULL};
+    struct ravl_node *node;
+    node = ravl_find(free_blocks, &data, RAVL_PREDICATE_EQUAL);
+    if (node) {
+        ravl_data_t *node_data = ravl_data(node);
+        if (node_data) {
+            head_node = node_data->value;
+        }
+    }
+    if (!head_node) {
+        head_node = malloc(sizeof(head_node_t));
+        if (!head_node) {
+            return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        }
+        head_node->head = NULL;
+        head_node->mutex = util_mutex_create();
+        ravl_data_t data = {(uintptr_t)block->size, head_node};
+        rv = ravl_emplace_copy(free_blocks, &data);
+        if (rv) {
+            return -1;
+        }
+    }
+
+    block->node = node_list_add(head_node, block);
+    if (!block->node) {
+        return -1;
+    }
+
+    assert(block->node->block->size == block->size);
+
+    return 0;
+}
+
+static block_t *free_blocks_get_ge(struct ravl *free_blocks, size_t size) {
```

Comment:
add a comment what this func does and when should be called

Reply:
Done